File: node.js/futures/signer.js
/**
 * MEXC Futures OpenAPI signer.
 * Note: Uses Futures signature rules, not Spot v3.
 */

const crypto = require('crypto')

/**
 * Filter null/undefined params.
 * @param {Object} params
 * @returns {Object}
 */
function filterParams (params) {
  if (!params || typeof params !== 'object') return {}
  const filtered = {}
  for (const [key, value] of Object.entries(params)) {
    if (value !== null && value !== undefined) {
      filtered[key] = value
    }
  }
  return filtered
}

/**
 * GET/DELETE: params sorted by key; key=value joined with &.
 * Raw values for signing (no URL encode).
 * @param {Object} params
 * @returns {string}
 */
function buildQueryStringForSign (params) {
  const filtered = filterParams(params)
  const keys = Object.keys(filtered).sort()
  return keys
    .map(k => `${k}=${filtered[k]}`)
    .join('&')
}

/**
 * POST: Use JSON string (no sorting). Array body is stringified directly.
 * @param {Object|Array} params
 * @returns {string}
 */
function buildJsonStringForSign (params) {
  if (Array.isArray(params)) return JSON.stringify(params)
  const filtered = filterParams(params)
  return JSON.stringify(filtered)
}

/**
 * Generate MEXC Futures signature.
 * @param {Object} options
 * @param {string} options.accessKey
 * @param {string} options.apiSecret
 * @param {string|number} options.requestTime - Millisecond timestamp
 * @param {string} options.method - GET | POST | DELETE
 * @param {Object} options.params - Business params (excluding path params)
 * @returns {string} Hex lowercase signature
 */
function sign (options) {
  const { accessKey, apiSecret, requestTime, method, params = {} } = options
  const reqTime = String(requestTime)
  let parameterString

  if (method === 'GET' || method === 'DELETE') {
    parameterString = buildQueryStringForSign(params)
  } else if (method === 'POST') {
    parameterString = buildJsonStringForSign(params)
  } else {
    parameterString = ''
  }

  const stringToSign = accessKey + reqTime + parameterString
  const hmac = crypto.createHmac('sha256', apiSecret)
  hmac.update(stringToSign)
  return hmac.digest('hex')
}

module.exports = {
  filterParams,
  buildQueryStringForSign,
  buildJsonStringForSign,
  sign
}

File: go/futures/mexcfutures/signer.go
package mexcfutures

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strconv"
	"time"
)

// Sign computes the MEXC Futures OPEN-API signature (HMAC-SHA256, lowercase hex).
// stringToSign = accessKey + requestTime + parameterString.
// GET/DELETE: parameterString is sorted query string (key=value&..., no URL encoding).
// POST: parameterString is the raw JSON string of the body (no key sorting; arrays preserved).
// See https://www.mexc.com/api-docs/futures/integration-guide.
func Sign(accessKey, apiSecret, requestTime, method string, params interface{}) (string, error) {
	if accessKey == "" || apiSecret == "" {
		return "", fmt.Errorf("accessKey and apiSecret required for sign")
	}
	reqTime := requestTime
	if reqTime == "" {
		reqTime = strconv.FormatInt(nowMs(), 10)
	}

	var parameterString string
	switch method {
	case "GET", "DELETE":
		m, ok := toMap(params)
		if !ok {
			parameterString = ""
		} else {
			filtered := FilterParams(m)
			parameterString = BuildQueryString(filtered)
		}
	case "POST":
		var err error
		parameterString, err = parameterStringForPost(params)
		if err != nil {
			return "", err
		}
	default:
		parameterString = ""
	}

	stringToSign := accessKey + reqTime + parameterString
	h := hmac.New(sha256.New, []byte(apiSecret))
	h.Write([]byte(stringToSign))
	sig := hex.EncodeToString(h.Sum(nil))
	return sig, nil
}

// parameterStringForPost returns the exact JSON string to sign for POST (no key sorting).
// Supports map or slice (raw array body).
func parameterStringForPost(params interface{}) (string, error) {
	if params == nil {
		return "", nil
	}
	switch v := params.(type) {
	case []interface{}:
		data, err := json.Marshal(v)
		if err != nil {
			return "", err
		}
		return string(data), nil
	case []byte:
		return string(v), nil
	}
	m, ok := toMap(params)
	if !ok {
		data, err := json.Marshal(params)
		if err != nil {
			return "", err
		}
		return string(data), nil
	}
	filtered := FilterParams(m)
	data, err := json.Marshal(filtered)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func toMap(params interface{}) (map[string]interface{}, bool) {
	if params == nil {
		return nil, false
	}
	m, ok := params.(map[string]interface{})
	return m, ok
}

// nowMs returns current Unix time in milliseconds (used when requestTime is empty).
var nowMs = func() int64 { return time.Now().UnixMilli() }
